spmd/tensor/device_mesh.py

```diff
@@ -16,48 +20,70 @@
     scatter,
 )
 
-# pyre-fixme[5]: Global expression must be annotated.
-_global_device_mesh = None
+_global_device_mesh: "DeviceMesh" = None
 
 
-# pyre-fixme[3]: Return type must be annotated.
-def get_global_device_mesh():
+def get_global_device_mesh() -> "DeviceMesh":
     global _global_device_mesh
     return _global_device_mesh
 
 
-# pyre-fixme[3]: Return type must be annotated.
-# pyre-fixme[2]: Parameter must be annotated.
-def set_global_device_mesh(mesh):
+def set_global_device_mesh(mesh: "DeviceMesh") -> None:
     global _global_device_mesh
     _global_device_mesh = mesh
 
 
 class DeviceMesh(object):
     """
-    Device Mesh object, can be used as a context manager.
-    By default describes the device ids, layout and serves
-    as a proxy for communication among the device lists.
+    DeviceMesh represents a mesh of devices, where layout of
+    devices could be represented as a n-d dimension array, and
+    each value of the n-d dimensional array is the global id
+    of the default process group ranks.
+
+    DeviceMesh could be used to describe the layout of devices
+    across the cluster, and serves as a proxy for communication
+    among the device lists within the cluster.
 
     We use the default ProcessGroup in this DeviceMesh class
```

Comment:
should we allow to pass a "parent" process group?

Reply:
We should probably always use default pg as parent group, there's bunch of reasons:
1. DeviceMesh is using logical device id in meshes, which is the default pg rank number, if we allow passing a parent group and use the parent group as id assignment, the ids between device meshes gonna be different and it's hard to tell the mapping
2. Currently in c10d now we are relying on some global id from default pg to do some communication, i.e. rank related c10d op like scatter, are using global rank as the src_rank, so we better keep the default pg with the global rank to do such operations.

We should allow people to slice the device mesh and get a sub device_mesh, which basically user could get the smaller sized group as they want, or user could construct a device mesh with only subset of ranks, to keep the logical id meaning in the mesh